### ISSUE TITLE
Pixel biasing in bidisperse system

### ISSUE THREAD

**Issue by zoeith** [soft-matter/trackpy#378]
Hi all,

Using the matlab version (http://site.physics.georgetown.edu/matlab/tutorial.html) with the following lines we get positions of both small and large particles with no apparent pixel biasing, but in the image attached, trackpy's locate function with similar input parameters has lots of pixel biasing.

```matlab
a = double(imread('00000.tif'));
b = bpass(a,1,9);
pk = pkfnd(b,5,9);
cnt = cntrd(b,pk,11);
```

I've tried varying the parameters including percentile and minmass, but I either end up missing many of the small particles or making the bias worse. I'm not familiar enough yet with the implementation details to know why this difference might show up, so hope someone here is.

[00000.tif.zip](https://github.com/soft-matter/trackpy/files/322495/00000.tif.zip)
(github doesn't support tif files :-/)

**Comment by caspervdw** [comment]
Are you using the numba or the python refine engine?

**Comment by caspervdw** [comment]
Because this could be an issue with the python engine, solved in #377. Please checkout that branch if you think that you are using the python engine.

**Comment by zoeith** [comment]
Cool, I getter better (but still weird) results with python:
![image](https://cloud.githubusercontent.com/assets/8890334/16179840/16bb069c-3640-11e6-9d41-f014b0cee3ca.png)
vs numba:
![image](https://cloud.githubusercontent.com/assets/8890334/16179845/33350d40-3640-11e6-92eb-60910c7baa5d.png)

**Comment by zoeith** [comment]
Trackpy 0.3.0
Numba version is 0.26.0

**Comment by caspervdw** [comment]
To be clear, the python and numba engines should do exactly the same. This will be fixed in v0.3.1. For now let's only consider the numba engine.

The subpixel biasing is (as far as I know) always caused by a too small mask size. But the difference with matlab is worrying, I thought that the output is always very similar. Maybe @nkeim has something to add in this? Which trackpy parameters did you use to compare with the matlab code that you posted earlier?

**Comment by zoeith** [comment]
The pixel biasing is pretty consistent with changes in parameters up to a diameter of 13 (but is still slightly present) when many of the small particles start to go missing.

An example function call might be:
```python
f = tp.locate(frames[0], 9.0, separation=9.5, minmass=0.0, noise_size=1.0,
              percentile=10, invert=False, characterize=False, engine='numba')
```

I've tried lots of various combinations of percentile, minmass, noise_size, and separation.

**Comment by caspervdw** [comment]
Did you try lower values of `separation`?

If that does not work, you'll probably have to wade into the `locate` code: more specifically, give `local_maxima` (which is the equivalent of pkfnd) a different length scale than `refine` (which is cntrd). We are planning on making this more user friendly, but for now the internals of locate are not properly documented. Please let us know your progress!

**Comment by zoeith** [comment]
Adding an optional `refine_diameter` parameter to `tp.locate` seems to have significantly improved the results!

```python
f = tp.locate(frames[0], 9.0, minmass=0.0, noise_size=1.0, separation=8.0, refine_diameter=14,
              percentile=10, invert=False, characterize=False, engine='numba',max_iterations=20)
```

![image](https://cloud.githubusercontent.com/assets/8890334/16210551/989b340a-370a-11e6-9f17-197d7739a38e.png)

I think I can propagate that through to `tp.batch` and generate some docs for the option if you want a PR.

Right now it just looks like this:
```python
    radius = tuple([x//2 for x in diameter])
    if refine_diameter is None:
        refine_radius = radius
    else:
        refine_diameter = validate_tuple(refine_diameter, ndim)
        refine_diameter = tuple([int(x) for x in refine_diameter])
        refine_radius = tuple([x//2 for x in refine_diameter])
```

**Comment by caspervdw** [comment]
Thanks for the report @zoeith , it looks much better!
I think that I am :-1:  on adding any extra kwargs to `locate`. (what do you think @danielballan  @tacaswell  @nkeim ?) We actually plan to reduce them and make it easier for users to write their own `locate` and `batch` for more specific cases. ( #257 )

But any contribution on trackpy is much appreciated, please let us know if you encounter an issue or if you have an enhancement.

**Comment by tacaswell** [comment]
I am actually :+1: on this (even though it expands the API a bit) in that the window used for refinement / COM calculations does not need to be the same value used for band passing.

I normally advocate for adding more specialized functions, but this seems like a reasonable generalization.

**Comment by zoeith** [comment]
I tried out #361 and got real improvement in the pixel biasing. But the diameter I used is still kind of large. It seems like most of the particles are there though.

```python
f = tp.locate(frames[0], 13, noise_size=1, separation=8.0, smoothing_size=7, minmass=0.0,
              percentile=10, invert=False, characterize=False, engine='numba')
```
![image](https://cloud.githubusercontent.com/assets/8890334/16276582/fc072616-387c-11e6-923f-15ed6d5886b6.png)


- It looks like all the components of `tp.bandpass()` are already exposed through the `tp.locate()` kwargs so I don't understand what turning off preprocess gains?
- I think adding noise to remove biasing is probably bad... but it might be preferable to non-physical unusually large jitter in trajectories? (If it actually does that.)

Thanks everyone!
